semgrep/semgrep/config_resolver.py

```diff
@@ -61,26 +64,165 @@
+        )
+        logger.debug(f"trying to download from {config_url}")
+        logger.info(downloading_msg.rstrip("\n"))
+
+        try:
+            config = parse_config_string(
+                "remote-url",
+                self._make_config_request(),
+                filename=f"{config_url[:20]}...",
+            )
+            logger.info(message_with_done(downloading_msg, overwrite=True))
+            return config
+        except Exception as e:
+            raise SemgrepError(f"Failed to download config from {config_url}: {str(e)}")
+
+    def _load_config_from_local_path(self) -> Dict[str, YamlTree]:
+        """
+        Return config file(s) as dictionary object
+        """
+        location = self.config_path
+        base_path = get_base_path()
+        loc = base_path.joinpath(location)
+
+        local_config_msg = f"Loading local config from {loc}..."
+        logger.info(local_config_msg)
+        if loc.exists():
+            if loc.is_file():
+                config = parse_config_at_path(loc)
+            elif loc.is_dir():
+                config = parse_config_folder(loc)
+            else:
+                raise SemgrepError(f"config location `{loc}` is not a file or folder!")
+        else:
+            addendum = ""
+            if IN_DOCKER:
+                addendum = " (since you are running in docker, you cannot specify arbitrary paths on the host; they must be mounted into the container)"
+            raise SemgrepError(
+                f"WARNING: unable to find a config; path `{loc}` does not exist{addendum}"
+            )
+        logger.info(message_with_done(local_config_msg, overwrite=True))
+        return config
+
+    def _make_config_request(self) -> str:
+        import requests  # here for faster startup times
+
+        config_url = self.config_path
+
+        headers = {"User-Agent": SEMGREP_USER_AGENT}
+        r = requests.get(config_url, stream=True, headers=headers, timeout=20)
+        if r.status_code == requests.codes.ok:
+            content_type = r.headers.get("Content-Type")
+            yaml_types = [
+                "text/plain",
+                "application/x-yaml",
+                "text/x-yaml",
+                "text/yaml",
+                "text/vnd.yaml",
+            ]
+            if content_type and any((ct in content_type for ct in yaml_types)):
+                return r.content.decode("utf-8", errors="replace")
+            else:
+                raise SemgrepError(
+                    f"unknown content-type: {content_type} returned by config url: {config_url}. Can not parse"
+                )
+        else:
+            raise SemgrepError(
+                f"bad status code: {r.status_code} returned by config url: {config_url}"
+            )
+
+    def is_registry_url(self) -> bool:
+        return self.from_registry
+
+    def __str__(self) -> str:
+        return self.config_path
+
+
 class Config:
```

Comment:
This is really for multiple Configs. Maybe `ConfigList`? `Configs`?

Reply:
This PR is only abstracting the types. This constructor if from earlier work. I'd prefer to keep refactors to a minimum here.

Reply:
Sounds good, I can refactor in another PR